src/main/java/org/scijava/table/DefaultTableIOPlugin.java

```diff
@@ -264,17 +226,70 @@ public GenericTable open(final String source) throws IOException {
 						" is not the same length as the first line.");
 				}
 				for (int i = 0; i < cols.size(); i++) {
-					table.set(i, lineNum - 1, parser.apply(cols.get(i)));
+					if(lineNum == 1 && readColHeaders) {
+						columnParsers.put(i, getParser(cols.get(i), i, options));
+					}
+					table.set(i, lineNum - 1, columnParsers.get(i).apply(cols.get(i)));
 				}
 			}
 		}
 		return table;
 	}
 
+	private static Function<String, Object> getParser(String content, int column, TableIOOptions.Values options) {
+		ColumnTableIOOptions.Values colOptions = options.column(column);
+		if(colOptions != null) return colOptions.parser();
+		if(options.guessParser()) return guessParser(content);
+		return options.parser();
+	}
+
+	static Function<String, Object> guessParser(String content) {
+		try {
+			Integer.valueOf(content);
+			return Integer::valueOf;
+		} catch(NumberFormatException ignored) {}
+		try {
+			Long.valueOf(content);
+			return Long::valueOf;
+		} catch(NumberFormatException ignored) {}
+		try {
+			Double.valueOf(content);
+			return Double::valueOf;
+		} catch(NumberFormatException ignored) {}
+		try {
+			Float.valueOf(content);
+			return Float::valueOf;
+		} catch(NumberFormatException ignored) {}
+		if(content.equalsIgnoreCase("true")||content.equalsIgnoreCase("false")) {
+			return Boolean::valueOf;
+		}
+		return String::valueOf;
+	}
+
 	@Override
 	public void save(final Table table, final String destination)
-		throws IOException
-	{
+			throws IOException {
+		TableIOOptions options = new TableIOOptions();
+		// DISCUSS if we guess at his point if column or row headers should be written, the plugin will behave differently
+		// if you call save(table, destination) vs. save(table, destination, new TableIOOptions())
+		// .. and this seems just wrong.
+//		options.writeColHeaders(table.getColumnCount() > 0 && //
+//				IntStream.range(0, table.getColumnCount()).allMatch(col -> table
+//						.getColumnHeader(col) != null));
+//		options.writeRowHeaders(table.getRowCount() > 0 && //
+//				IntStream.range(0, table.getRowCount()).allMatch(row -> table
+//						.getRowHeader(row) != null));
```

Comment:
I agree we should aim for predictable and deterministic behavior. Guessing whether to write row and/or column headers should be opt-in (if at all).